File: backend/gamereview.py
import chess
import io
import chess.pgn
from stockfish import Stockfish
from classifymoves import classifyMoves, countMoveCategories


def getEngineAnalysis(FENs):
    stockfish = Stockfish(path=r"stockfish\stockfish-windows-x86-64-avx2.exe", depth=16, parameters={"Threads": 1, "Minimum Thinking Time": 1, "Hash": 32, "Slow Mover": 100})
    response = []
    for count, FEN in enumerate(FENs):
        if (count == 0):
            bestmove = False
        else:
            stockfish.set_fen_position(FENs[count-1])
            bestmove = stockfish.get_best_move()   

        stockfish.set_fen_position(FEN)
        eval = stockfish.get_evaluation()
        compiled = {'move_no': count, 'fen': FEN, 'best_move': bestmove, 'eval': eval}
        response.append(compiled)
        print(count)

    return response

def changeFormat(pgn, infos, moves):
    game = chess.pgn.read_game(io.StringIO(pgn))
    board = game.board()
    default_fen = board.fen()
    for counter, (info, move) in enumerate(zip(infos, moves)):

        if (info['eval']['type']!='mate'):
            info['eval']['value']/=100
        if (not info['best_move']):
            info['best_move'] = (None)
        else:
            board.set_fen(infos[counter-1]['fen'])
            info['best_move'] = (board.san(chess.Move.from_uci(info['best_move'])))
            board.set_fen(default_fen)    
        info['move'] = move
    return infos

def review_game (pgn):
    game = chess.pgn.read_game(io.StringIO(pgn))
    board = game.board()
    moves = game.mainline_moves()
    moves_fens = [board.fen()]
    moves_san = [None]

    for move in moves:
        moves_san.append(board.san_and_push(move))
        moves_fens.append(board.fen())

    analysis = getEngineAnalysis(moves_fens)
    analysis = changeFormat(pgn, analysis, moves_san)
    analysis = classifyMoves(analysis)
    analysis = countMoveCategories(analysis, pgn)
    return analysis

File: src/scripts/gamereview.js
import validatePGN from "./validation.js";
import gameLoaded from "../assets/sound/game-loaded.mp3";
import showErrorMessage from "./errorMessage";
import classifyMoves, { countMoveCategories } from "./classifymoves.js";
import getEngineAnalysis from "./engine.js";
import { Chess } from "chess.js";
import getGames from "./getgames.js";

const review_game = async (
  input,
  type,
  setPGN,
  month,
  year,
  setGames,
  setIsOpen,
  setUsername
) => {
  switch (type) {
    case "pgn": {
      const valid = validatePGN(input);
      if (valid) {
        await analyse(input, setPGN);
      } else {
        showErrorMessage("Invalid PGN");
      }
      break;
    }
    case "chess.com": {
      const games = await getGames(input, month, year);
      setUsername(input);
      setIsOpen(true);
      setGames(games);
      break;
    }
  }
};

export const analyse = async (input, setPGN) => {
  let depth = 5;
  try {
    const FENs = getFENs(input);
    let analysis = await getEngineAnalysis(FENs, depth);
    let moves_san = [null];

    const tempChess = new Chess();
    tempChess.loadPgn(input);
    moves_san = moves_san.concat(tempChess.history());

    analysis = changeFormat(input, analysis, moves_san);
    analysis = classifyMoves(analysis);
    analysis = countMoveCategories(analysis, input);

    setPGN(analysis);
    const sound = new Audio(gameLoaded);
    sound.play();
  } catch (e) {
    showErrorMessage(e);
  }
};

const getFENs = (pgn) => {
  const chess = new Chess();
  chess.loadPgn(pgn);
  const moves = chess.history();
  const numberOfMoves = moves.length;
  chess.reset();
  const FENs = [chess.fen()];
  for (let x = 0; x <= numberOfMoves - 1; x++) {
    chess.move(moves[x]);
    FENs.push(chess.fen());
  }
  return FENs;
};

const changeFormat = (pgn, infos, moves) => {
  const chess = new Chess();
  chess.loadPgn(pgn);
  try {
    for (let counter = 0; counter < infos.length; counter++) {
      const info = infos[counter];
      const move = moves[counter];
      if (info.eval.type !== "mate") {
        info.eval.value /= 100;
      }
      if (!info.best_move) {
        info.best_move = null;
      }
      info.move = move;
    }
  } catch (e) {
    showErrorMessage(e);
  }
  return infos;
};

export default review_game;
